typer-cli-python/source_code_step_6/tests/test_rptodo.py

```diff
@@ -0,0 +1,125 @@
+import json
+
+import pytest
+from typer.testing import CliRunner
+
+from rptodo import (
+    DB_READ_ERROR,
+    ID_ERROR,
+    SUCCESS,
+    __app_name__,
+    __version__,
+    cli,
+    rptodo,
+)
+
+runner = CliRunner()
+
+
+def test_version():
+    result = runner.invoke(cli.app, ["--version"])
+    assert result.exit_code == 0
+    assert f"{__app_name__} v{__version__}\n" in result.stdout
+
+
+@pytest.fixture
+def mock_json_file(tmp_path):
+    todo = [{"Description": "Get some milk.", "Priority": 2, "Done": False}]
+    db_file = tmp_path / "todo.json"
+    with db_file.open("w") as db:
+        json.dump(todo, db, indent=4)
+    return db_file
+
+
+test_data1 = {
+    "description": ["Clean", "the", "house"],
+    "priority": 1,
+    "todo": {
+        "Description": "Clean the house.",
+        "Priority": 1,
+        "Done": False,
+    },
+}
+test_data2 = {
+    "description": ["Wash the car"],
+    "priority": 2,
+    "todo": {
+        "Description": "Wash the car.",
+        "Priority": 2,
+        "Done": False,
+    },
+}
+
+
+@pytest.mark.parametrize(
+    "description, priority, expected",
+    [
+        pytest.param(
+            test_data1["description"],
+            test_data1["priority"],
+            (test_data1["todo"], SUCCESS),
+        ),
+        pytest.param(
+            test_data2["description"],
+            test_data2["priority"],
+            (test_data2["todo"], SUCCESS),
+        ),
+    ],
+)
+def test_add(mock_json_file, description, priority, expected):
+    todoer = rptodo.Todoer(mock_json_file)
+    assert todoer.add(description, priority) == expected
+    read = todoer._db_handler.read_todos()
+    assert len(read.todo_list) == 2
+
+
+@pytest.fixture
+def mock_wrong_json_file(tmp_path):
+    db_file = tmp_path / "todo.json"
+    return db_file
+
+
+def test_add_wrong_json_file(mock_wrong_json_file):
+    todoer = rptodo.Todoer(mock_wrong_json_file)
+    response = todoer.add(["test task"], 1)
+    assert response.error == DB_READ_ERROR
+    read = todoer._db_handler.read_todos()
+    assert len(read.todo_list) == 0
+
+
+@pytest.fixture
+def mock_wrong_json_format(tmp_path):
+    db_file = tmp_path / "todo.json"
+    with db_file.open("w") as db:
+        db.write("")
+    return db_file
+
+
+def test_add_wrong_json_format(mock_wrong_json_format):
+    todoer = rptodo.Todoer(mock_wrong_json_format)
+    assert todoer.add(test_data1["description"], test_data1["priority"]) == (
+        test_data1["todo"],
+        SUCCESS,
+    )
+    read = todoer._db_handler.read_todos()
+    assert len(read.todo_list) == 1
+
+
+test_todo1 = {
+    "Description": "buy some milk.",
```

Comment:
⌨️ **Typo:** The test involving this data fails because it expects "Get some milk.":

```
(venv) ➜  source_code_step_6 git:(typer-cli-python) ✗ python -m pytest tests/    
==================================================== test session starts ====================================================
platform darwin -- Python 3.10.0b1, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /Users/martin/Documents/realpython/materials/typer-cli-python/source_code_step_6
collected 7 items                                                                                                           

tests/test_rptodo.py .....F.                                                                                          [100%]

========================================================= FAILURES ==========================================================
________________________________________________ test_set_done[1-expected0] _________________________________________________

mock_json_file = PosixPath('/private/var/folders/52/0zsd69115l57tf4st_1fw69w0000gn/T/pytest-of-martin/pytest-4/test_set_done_1_expected0_0/todo.json')
todo_id = 1, expected = ({'Description': 'buy some milk.', 'Done': True, 'Priority': 2}, 0)

    @pytest.mark.parametrize(
        "todo_id, expected",
        [
            pytest.param(1, (test_todo1, SUCCESS)),
            pytest.param(3, (test_todo2, ID_ERROR)),
        ],
    )
    def test_set_done(mock_json_file, todo_id, expected):
        todoer = rptodo.Todoer(mock_json_file)
>       assert todoer.set_done(todo_id) == expected
E       AssertionError: assert CurrentTodo(t...rue}, error=0) == ({'Descriptio...ority': 2}, 0)
E         At index 0 diff: {'Description': 'Get some milk.', 'Priority': 2, 'Done': True} != {'Description': 'buy some milk.', 'Priority': 2, 'Done': True}
E         Use -v to get the full diff

tests/test_rptodo.py:125: AssertionError
================================================== short test summary info ==================================================
FAILED tests/test_rptodo.py::test_set_done[1-expected0] - AssertionError: assert CurrentTodo(t...rue}, error=0) == ({'Desc...
================================================ 1 failed, 6 passed in 0.04s ================================================
```

**Note:** The same issue also persists in `source_code_step_7` (but it's fixed in `source_code_final`). Please fix it for the code in step 7 as well.